### ISSUE TITLE
Add BufferExamplePath class 

### ISSUE THREAD

**Issue by capital-G** [supercollider/supercollider#5859]
## Motivation

*Forked* from https://github.com/supercollider/supercollider/issues/5604 which discusses adding more included samples to SC.

```supercollider
b = Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01-44_1.aiff");
```

is often used for examples or in teaching environments as it is the only one of two included samples in SC .
Both samples do not have a trivial filenames, making it necessary to copy these lines all the time (which are hidden in examples or in the `Buffer` documentation).
It also lacks any kind of documentation what those samples are.

## Description of Proposed Feature

Adding a new Class which abstracts those file paths makes accessing those examples more easy via auto-completion and dedicated documentation, also allowing for an interface to add new examples (see #5604 ).

Python libraries also handle this via namespaces/classes, see https://github.com/supercollider/supercollider/issues/5604#issuecomment-1232926043

## Plan for Implementation

Implementation is rather trivial, it could look something like

```supercollider
BufferExamplePaths {
  *nasa { ^Platform.resourceDir +/+ "sounds/a11wlk01-44_1.aiff" }

  *baby { ^Platform.resourceDir +/+ "sounds/a11wlk01.wav" }
}
```

replacing

```supercollider
b = Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01-44_1.aiff");
```

with

```supercollider
b = Buffer.read(s, BufferExamplePaths.nasa);
```

**Comment by dyfer** [comment]
I'd prefer to not have "buffer" in the name, but rather SoundFilePaths or SoundFileExamplePaths or SoundExamplePaths. However, I'd like to hear what others think so please don't make changes just yet.

**Comment by capital-G** [comment]
The motivation to use/start with `Buffer` in the name is to find it easily when just typing `Buffer` into the IDE.

**Comment by dyfer** [comment]
I don't think it should be called Buffer if it doesn't return a buffer. 
We could also consider names like ExampleSounds etc...

**Comment by redFrik** [comment]
one issue i have with this is that for complete beginners it might be unintuitive to replace `BufferExamplePaths.nasa` with a custom string path.
the current `Platform.resourceDir +/+ "sounds/a11wlk01-44_1.aiff"` at least hint at what's required.

**Comment by telephon** [comment]
Yes, I now see that it is actually less useful for examples than for experimentation that is still cross-platform compatible.

**Comment by capital-G** [comment]
> I don't think it should be called Buffer if it doesn't return a buffer.

`SoundFile` seems to be indeed a more fitting name. I must admit I was not aware of this class before.
Maybe use an additional constructor for SoundFile like

```supercollider
Buffer.read(server: s, path: SoundFile.nasa.path);
```

Seems like a clean programming style and also clear what happens(?), at least to me.
The only downside IMO is that it seems to advocate a necessity of `SoundFile`.


> the current `Platform.resourceDir +/+ "sounds/a11wlk01-44_1.aiff"` at least hint at what's required.

IMO this can be better indicated by writing out the name of the arguments (which I am a strong advocate of), e.g. the statement above is to me clearer than the often seen

```supercollider
Buffer.read(s, Platform.resourceDir +/+ "sounds/a11wlk01-44_1.aiff"); 
```

Not just because of the arguments but it also introduces `+/+` (which is Path exclusive and maybe confuse the usage of strings in general?) and I need to understand `Platform.resourceDir` - resourceDir may be obvious if you speak english, but you need to understand what a `Platform` is (also in the last case one must remind oneself what `s` stands for).

**Comment by jamshark70** [comment]
In #5604 I had also suggested that a library of soundfile paths needs a design process to determine requirements and avoid the risk of adding something which would need to be deprecated later.

Predictably, an attempt to push through a class *without* a design process has resulted in disagreement over the ideal interface.

My suggestion is to RFC it, gather requirements, and write code *after* we have a clear idea what we need.

**Comment by capital-G** [comment]
I don't agree on that a NFC is necessary for this: It does not interfere, modify or deprecate any existing code and it is not a major change, only adding convenience/syntax sugar for demo/teaching purposes which is geared more towards novices (or even for me as advanced user b/c I can not remember those paths and need to copy them from the docs every time). Therefore I think the motivation is given, and having as much auto-completion should be a requirement as this makes it less prone to looking something up in the documentation again.
Where this functionality can be put precisely can be discussed in an/this issue IMO.

To me the most intuitive way would be to have a class/namespace which returns an actual Buffer but as the garbage collection of the language does not match the one on the server this is probably not a good idea as this can clog up the server memory. Always returning the *same* Buffer on the server is also not an option as a Buffer is mutable.

As there seems to be opinions against creating a new class for the proposed example management we need to put it into existing classes (probably as additional constructor?) where the options are (please extend, I am not too familiar with all the existing SC classes)

Class | Argument
--- | ---
`Buffer` | Most intuitively fitting ("i simply want any example buffer please") but b/c of SC inner workings (see above) probably not a good place
`SoundFile` | After 3 years of SC I was not aware of this class, but seems quite fitting as we want to give quick access to such a given file. It would lead to something like `Buffer.read(server: s, path: SoundFile.nasa.path)` which to me looks strange b/c it introduces `SoundFile` (which is also rarely (?) used), suggesting it is a necessary class when working with wav files and else. Also [the documentation is already quite big](https://doc.sccode.org/Classes/SoundFile.html), making it hard to directly find the proposed annotations for the examples.

After putting much thought into it for me the first draft creating a `BufferExamplePath` class seems most clean and intuitive to me:
* Following the single responsibility principle which leads to
* having a clean/simple documentation,
* it pops up when someone writes down `Buffer` in the IDE, leading to easy adaption
* yet its formulation is not misleading and can be easily understood in the context of *older* examples using `Platform.resourceDir`

**Comment by capital-G** [comment]
Coming back to this, I think the initial proposition would still be a good addition and would still like to give it one more try. I now added the class to to my installation/quark

```supercollider
SoundFileExamplesPath {
	*huston {
		^Platform.resourceDir +/+ "sounds" +/+ "a11wlk01-44_1.aiff";
	}

	*sinedPink {
		^Platform.resourceDir +/+ "sounds" +/+ "SinedPink.aiff";
	}

	*girl {
		^Platform.resourceDir +/+ "sounds" +/+ "a11wlk01.wav";
	}
}

```

IMO this is especially targeted towards novices, but even I as an advanced users would love to type less and not remember `a11wlk01.wav` and `a11wlk01-44_1.aiff` - I always need to look the proper path up in the `Buffer` docs.
I think also a video speaks for itself, which already omits the look up procedure

https://github.com/supercollider/supercollider/assets/8267062/e2f07a0b-4def-49f5-9f4d-82e586ce8fae

Addressing some concerns from my POV

Argument | Response
--- | ---
Make this a dynamic sample library | I think this is beyond the proposed scope and would not allow for autocomplete (which is the whole intention here). This would be a good addition via a Quark?
This obfuscates that `Buffer.read` wants a string as a path | I agree, but `Platform.resourceDir +/+ "sounds" +/+ "a11wlk01.wav"` also obfuscates that it is indeed a String as `Platform.resourceDir` could be anything. Also the `+/+` syntax is still a mysterious/advanced thing, at least to me. In order to make it obvious that it is a path it needs to be something along `/Applications/SuperCollider.app/Contents/Resources/sounds/a11wlk01.wav` which would be even worse to type/remember and non-cross-plattform compatible. Also why is it `Platform.resourceDir +/+ "sounds/a11wlk01.wav"` instead of `Platform.resourceDir +/+ "sounds" +/+ "a11wlk01.wav"` and why is `Platform.resourceDir` returning a 
A class which just returns static paths is bloat | Yes, but as SC is OO-only, there is no way to store these paths elsewhere. Adding it to an already existing class would most likely create bloat and it is not clear to which it should be added? `String` because it returns a string?
Use the actual filenames | I think this defeats the purpose of something easy to access. Also`a11wlk01-44_1.aiff` could also not be auto comlpeted.
Add other media such as pictures as well | I think this is also beyond scope and makes the interface less clear? But we could also adapt this by implementing something like `ExampleFiles.nasa` 

Some problems I see in the sketched code above

* `SoundFileExamplesPath` is suggesting that this is somehow related to `SoundFile` - it is not
* `SoundFileExamplesPath` is suggesting that it returns a `Path` (not existing in sclang) or `PathName` - it is not as `Buffer.read` is not accepting a `PathName`
* Is `.girl` a fitting name or should it be named something differently? The abstractness of this sample and lack of meaningful filename makes it difficult. Maybe there are some more creative or sensitive people around here which can make a better proposition?

The advantages:

* Easy autocomplete
* We could document something about these samples - are they free to use, what are they actually, what different example samples are there?
* In the future it would allow us to add new examples in a clear manner.

I'd personally switch out every ocassion of `Platform.resourceDir +/+ "sounds" +/+ "a11wlk01.wav"` with `SoundFileExamplePath.huston` in the docs as otherwise the class would not be "popular" and would not serve its purpose of adaptation.
But I already know that this would be controversial, yet I think this would make live of everyone easier - especially beginners.
Additionally, this would also not deprecate anything.

**Comment by telephon** [comment]
> * Is `.girl` a fitting name or should it be named something differently? The abstractness of this sample and lack of meaningful filename makes it difficult. Maybe there are some more creative or sensitive people around here which can make a better proposition?

It is not a girl, I think. Call it "voice", that feels more respectful?

**Comment by cdbzb** [comment]
.child ?Sent from my iPhoneOn May 31, 2024, at 11:13 AM, Julian Rohrhuber ***@***.***> wrote:﻿


Is .girl a fitting name or should it be named something differently? The abstractness of this sample and lack of meaningful filename makes it difficult. Maybe there are some more creative or sensitive people around here which can make a better proposition?


It is not a girl, I think. Call it "voice", that feels more respectful?
